File: solutions/python/trees/all-nodes-distance-k-in-binary-tree.py
# LeetCode 863 — All Nodes Distance K in Binary Tree (Medium)
# Category: Trees · Approach: Parent map + BFS
# Time: O(n) | Space: O(n)
# Source: https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/

from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val, self.left, self.right = val, left, right

def distance_k(root: TreeNode, target: int, k: int) -> list[int]:
    parent: dict[TreeNode, TreeNode | None] = {}
    target_node = None

    def dfs(node, par):
        nonlocal target_node
        if not node:
            return
        parent[node] = par
        if node.val == target:
            target_node = node
        dfs(node.left, node)
        dfs(node.right, node)

    dfs(root, None)
    if target_node is None:
        return []
    seen = {target_node}
    q = deque([target_node])
    dist = 0
    while q and dist < k:
        for _ in range(len(q)):
            node = q.popleft()
            for nb in (node.left, node.right, parent[node]):
                if nb and nb not in seen:
                    seen.add(nb)
                    q.append(nb)
        dist += 1
    return [node.val for node in q] if dist == k else []

File: solutions/typescript/trees/all-nodes-distance-k-in-binary-tree.ts
// LeetCode 863 — All Nodes Distance K in Binary Tree (Medium)
// Category: Trees · Approach: Parent map + BFS
// Time: O(n) | Space: O(n)
// Source: https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/

class TreeNode {
  val: number;
  left: TreeNode | null = null;
  right: TreeNode | null = null;
  constructor(val = 0) {
    this.val = val;
  }
}

function distanceK(root: TreeNode | null, target: number, k: number): number[] {
  const parent = new Map<TreeNode, TreeNode | null>();
  let targetNode: TreeNode | null = null;
  const dfs = (node: TreeNode | null, par: TreeNode | null) => {
    if (!node) return;
    parent.set(node, par);
    if (node.val === target) targetNode = node;
    dfs(node.left, node);
    dfs(node.right, node);
  };
  dfs(root, null);
  if (!targetNode) return [];
  const seen = new Set<TreeNode>([targetNode]);
  let frontier: TreeNode[] = [targetNode];
  let dist = 0;
  while (frontier.length && dist < k) {
    const next: TreeNode[] = [];
    for (const node of frontier) {
      for (const nb of [node.left, node.right, parent.get(node)!]) {
        if (nb && !seen.has(nb)) {
          seen.add(nb);
          next.push(nb);
        }
      }
    }
    frontier = next;
    dist++;
  }
  return dist === k ? frontier.map((n) => n.val) : [];
}
